src/entity_extraction/training/spacy_ner/spacy_preprocess.py

```diff
@@ -0,0 +1,266 @@
+import logging
+import shutil
+import glob
+from docopt import docopt
+from spacy.tokens import DocBin
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+opt = docopt(__doc__)
+
+def preprocess_data():
+    # If the data_path consists of JSON, load all the files in the directory
+    nlp = spacy.blank("en")
+    # Creating artifacts for only train and val sets
+    train_doc_bin, val_doc_bin = DocBin(), DocBin()
+    
+    # Segragating all files for evaluation purposes
+    train_data = {}
+    val_data = {}
+    test_data = {}
+    
+    train_path = os.path.join(opt['--data_path'], "train")
+    val_path = os.path.join(opt['--data_path'], "val")
+    test_path = os.path.join(opt['--data_path'], "test")
+
+    if os.path.exists(train_path):
+        shutil.rmtree(train_path)
+        logger.info(f"The folder '{train_path}' has been deleted.")
+    if os.path.exists(val_path):
+        shutil.rmtree(val_path)
+        logger.info(f"The folder '{val_path}' has been deleted.")
+    if os.path.exists(test_path):
+        shutil.rmtree(test_path)
+        logger.info(f"The folder '{test_path}' has been deleted.")
+
+    files = glob.glob(os.path.join(opt['--data_path'], "*"))
+    train_ids, val_ids, test_ids = split_train_val_test(opt['--data_path'],
+                                                        float(opt['--train_split']),
+                                                        float(opt['--val_split']),
+                                                        float(opt['--test_split']))
+    
+    for labelled_file in files:
+        entities = []
+        with open(labelled_file, 'r') as fin:
+            article = fin.readlines()
+            article_data = json.loads(article[0])
+            text = article_data['task']['data']["text"]
+            gdd_id = article_data['task']['data']["gdd_id"]
+        
+        doc = nlp.make_doc(text)
+
+        for label in article_data['result']:
+            start = label['value']['start']
+            end = label['value']['end']
+            ent = label['value']['labels'][0]
+            span = doc.char_span(start, end, label=ent)
+            if span is not None:
+                entities.append(span)
+                
+        doc.ents = entities
+        
+        if gdd_id in train_ids:
+            train_doc_bin.add(doc)
+            train_data[f'{labelled_file.split("/")[-1]}'] = article_data
+        elif gdd_id in val_ids:
+            val_doc_bin.add(doc)
+            val_data[f'{labelled_file.split("/")[-1]}'] = article_data
+        else:
+            test_data[f'{labelled_file.split("/")[-1]}'] = article_data
+    
+    
+    
+    # Write the artifacts to the same data directory
+    os.makedirs(train_path, exist_ok=True)
+    os.makedirs(val_path, exist_ok=True)
+    os.makedirs(test_path, exist_ok=True)
+
+    train_doc_bin.to_disk(os.path.join(train_path, "train.spacy"))
+    val_doc_bin.to_disk(os.path.join(val_path, "val.spacy"))
```

Comment:
bug: where is test.spacy created?

Reply:
`test.spacy` artifact is not used during training hence it is not required. During the evaluation, we are using the JSON files, thus I didn't create it.